### PR TITLE
[2.1] - Add command to view mview state and queue

### PR DESCRIPTION
**Pull request by convenient**
- PR for 2.1 = #12050
- PR for 2.2 = #12122
- PR for 2.3 = #12592

### Description

This is similar to the magerun1 command here: https://github.com/netz98/n98-magerun/pull/891

I like the ability to view the mview queue in realtime as its being processed, it can be quite helpful when debugging indexing issues.

This command will actually show how many items are in the list pending processing, as well information from the `mview_state` table.

```
php bin/magento indexer:status:mview
+---------------------------+----------+--------+---------------------+------------+---------+
| ID                        | Mode     | Status | Updated             | Version ID | Backlog |
+---------------------------+----------+--------+---------------------+------------+---------+
| catalog_category_product  | enabled  | idle   | 2017-11-02 10:00:00 | 1          | 0       |
| catalog_product_attribute | enabled  | idle   | 2017-11-02 10:00:00 | 1          | 1       |
| catalog_product_category  | disabled | idle   | 2017-11-02 10:00:00 | 1          | 0       |
| catalog_product_price     | enabled  | idle   | 2017-11-02 10:00:00 | 1          | 0       |
+---------------------------+----------+--------+---------------------+------------+---------+
```

screen shot 2017-11-05 at 20 47 47

### Manual testing scenarios
1. Enable an indexer `php bin/magento indexer:set-mode schedule catalog_product_price`
2. Trigger an action on the product price that will update the `_cl` table
3. Run `php bin/magento indexer:status:mview` to see the backlog increase
4. `magerun2 sys:cron:run indexer_update_all_views` will trigger the mview processes
5. See the backlog decrease 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
Really good idea. Would be great to have something similar in the magento admin.

**Comment by convenient** [comment]
> Would be great to have something similar in the magento admin

@ihor-sviziev That would definitely be neat, however I'll focus on getting this command and potentially deal with admin UI in a separate PR at a later.

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * Command for displaying status of mview indexers.
+ */
+class IndexerStatusMviewCommand extends Command
+{
+    /** @var \Magento\Framework\Mview\View\CollectionInterface $mviewIndexersCollection */
+    private $mviewIndexersCollection;
+
+    public function __construct(
+        \Magento\Framework\Mview\View\CollectionInterface $collection
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: You need to use \Magento\Framework\Mview\View\CollectionFactory there. Also would be great to import this class.

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+                $pendingCount = count($changelog->getList($state->getVersionId(), $currentVersionId));
+
+                $pendingString = "<error>$pendingCount</error>";
+                if ($pendingCount <= 0) {
+                    $pendingString = "<info>$pendingCount</info>";
+                }
+
+                $rows[] = [
+                    $indexer->getData('view_id'),
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: Would be better to use $indexer->getId()

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        try {
+            $table = $this->getHelperSet()->get('table');
+            $table->setHeaders(['ID', 'Mode', 'Status', 'Updated', 'Version ID', 'Backlog']);
+
+            $rows = [];
+
+            /** @var \Magento\Framework\Mview\View $indexer */
+            foreach ($this->mviewIndexersCollection as $indexer) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: Actually it's not indexer, it's view. So it's better to use one name for this object type

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                $pendingCount = count($changelog->getList($state->getVersionId(), $currentVersionId));
+
+                $pendingString = "<error>$pendingCount</error>";
+                if ($pendingCount <= 0) {
+                    $pendingString = "<info>$pendingCount</info>";
+                }
+
+                $rows[] = [
+                    $indexer->getData('view_id'),
+                    $state->getData('mode'),
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: $state->getMode()

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    $state->getData('status'),
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: $state->getStatus()

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                    $state->getData('updated'),
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: $state->getUpdated()

File: app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                    $state->getData('version_id'),
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Console/Command/IndexerStatusMviewCommand.php]: $state->getVersionId()

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by convenient** [comment]
Thanks for the review @ihor-sviziev I'll action these later. 

Quick Q: why is it your preference to use `$obj->getFoo()` rather than `$obj->getData('foo')` ? I've a personal dislike for those magic methods but it'd just a style thing for me. Is this to keep it more inline with the rest of M2?

**Comment by ihor-sviziev** [comment]
You're looking on implementation. If you look on interface - you'll see
that there is no method getData, but has all methods that I wrote you.

**Comment by convenient** [comment]
@ihor-sviziev thank you for the clarification, you were right I was not considering the interface.

i've updated the PR to address your concerns and the test suite is green, please review.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by convenient** [comment]
Cool, now that this is going green I'll port it to 2.2.

**Comment by convenient** [comment]
Done. Raised #12122 for the 2.2 series.

**Comment by convenient** [comment]
FYI I'll not do any further work on this PR until everyone's happy with the changes and discussion that will be happening in #12122 .

**Pull request merged**